processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -2096,7 +2096,16 @@ private boolean doesNotNeedPresenceCheckForSourceParameter(PropertyMapping mappi
             return false;
         }
 
-        return mapping.getAssignment().isSourceReferenceParameter();
+        return mapping.getAssignment().isSourceReferenceParameter()
+            || isCheckedByGuardClause( mapping );
+    }
+
+    private boolean isCheckedByGuardClause(PropertyMapping mapping) {
+        boolean isUpdateMethod = getReturnTypeToConstruct() == null && isMapNullToDefault();
+        return !isUpdateMethod
+            && presenceChecksByParameter.size() == 1
+            && presenceChecksByParameter.containsKey( mapping.getSourceBeanName() )
+            && mapping.getAssignment().getSourcePresenceCheckerReference() == null;
```

Comment:
Why is this needed? All tests pass without this, did I remove some edge case in the existing test cases?

Reply:
Are you refering to only the last condition? 
`mapping.getAssignment().getSourcePresenceCheckerReference() == null`

Reply:
Yes indeed, only the last condition

Reply:
That is indeed a really good question.
I'm going to rebase and try a little if I can find out why I thought I need it.
If I can reproduce it, I'm going to remove the condition and update the PR.

Reply:
Please also check if the class and methods I removed from the tests are causing everything to be green

Reply:
I checked out the older revision and none of my Testcases covered that line. 
Sadly I can't tell you my initial thought when I wrote that line.